### ISSUE TITLE
Copy applications

### ISSUE THREAD

**Issue by keith** [caskroom/homebrew-cask#3888]
Based on [a conversation](https://news.ycombinator.com/item?id=6999615) on HackerNews a while back I was looking into the codebase to see about implementing the ability to copy applications rather than symlinking them. As I argued then, since externally installed applications often have their own upgrade mechanisms I believe it would make sense for cask to provide users with the ability to download, unpack and copy applications to the target directory. This would mean that cask would have no knowledge of the application after it was installed and it would not show up in the `list`. For all intensive purposes to cask it would be uninstalled. Yet this way the application would live on in the installed location updating itself internally as needed. While this may seemingly not entirely utilize cask's feature set I still think the download and unpacking process would be extremely useful regardless of what happens to the application after that.

I would love to hear thoughts about adding this feature. As for how that would be done, while briefly looking through the codebase I think I traced the installation process to the artifact installers depending on the source's type. It seems to me that this would present a problem. In the example of an application, since the app artifact [inherits directly from the symlink installer](https://github.com/caskroom/homebrew-cask/blob/master/lib/hbc/artifact/app.rb), it would take some other changes to make copying work cleanly. Although another, much less clean option, would be to provide an option that changes the [symlink command](https://github.com/caskroom/homebrew-cask/blob/master/lib/hbc/artifact/symlinked.rb#L12) since simply changing it to a `cp` with the correct arguments nearly accomplishes what I'm suggesting (although it obviously wouldn't remove the original application probably in `/opt/...`).

**Comment by budda** [comment]
:+1:  Would love to use Cask for simply copying a downloaded app in to /Applications -- set by an ENV variable for now would make me happy with Ansible :)

**Comment by keith** [comment]
I've created a [homebrew formula](https://github.com/keith/homebrew-formulae/blob/master/Formula/brew-cask.rb) with a patch that just switches `ln` -> `cp` as a temporary fix.

**Comment by szhu** [comment]
@keith can you give an example of how to install and use this formula? Thanks!

**Comment by keith** [comment]
```
brew tap caskroom/cask; brew install keith/formulae/brew-cask 
```

**Comment by szhu** [comment]
Thanks! btw it seems to not work with casks that require command-line tools, is it because the source is being deleted too early?

```
==> Downloading https://central.github.com/mac/latest
######################################################################## 100.0%
==> Copying App 'GitHub.app' to '/Users/Administrator/Applications/GitHub.app'
==> Running uninstall process for github; your password may be necessary
==> Removing launchctl service com.github.GitHub.Conduit
==> Removing launchctl service com.github.GitHub.GHInstallCLI
Error: It seems the copy source is not there: '/opt/homebrew-cask/Caskroom/github/latest/GitHub.app/Contents/MacOS/github_cli'
```

**Comment by keith** [comment]
Ah interesting. I didn't test with GitHub for mac. Looks like some things are suppose to get copied after I removed the source. In theory I don't have to remove the source, it's just unnecessary cruft. An immediate fix to this would be to remove the one you just installed and install my previous revision.

```
brew rm brew-cask
brew install https://raw.githubusercontent.com/keith/homebrew-formulae/ace116da352b6cc88d19b574a000af011017a2ed/Formula/brew-cask.rb
```

This one should solve that problem but the source will sit around in `/opt/homebrew-cask/...` forever.

**Comment by szhu** [comment]
Thanks for being on top of this! I think I'll stick with standard brew-cask for now. Would it be possible to change the patch so that it removes everything only after everything else is applied?

**Comment by keith** [comment]
Possibly.

**Comment by danielbayley** [comment]
+1 for this, except I think move rather than copy since we don't want 2 of everything taking up unnecessary space.

At the moment I have a Hazel rule watching /Applications which swaps the link with the App like this;
```bash
app=$(readlink "$1")
rm "$1" && mv "$app" /App*s && ln -s /App*s/"${app##*/}" "${app%/*}"
```

**Comment by vitorgalvao** [comment]
Closing in favour of these news: #13201.

**Comment by danielbayley** [comment]
In the mean time, if any other [Hazel](http://noodlesoft.com) users (@brianroach) want to use my rule to do this, I've uploaded it [here](http://cl.ly/dSha). Just import the rules and apply them to
`/usr/local/opt/Caskroom` (or wherever your `Caskroom` folder is…)

The main chunk of [bash] code is just;
```bash
version() { mdls -name kMDItemVersion "$1"| egrep -o [0-9.]+ ;}
App=(/App*s/"${1##*/}")

[[ ! -a $App || -L $App || $(version $1) > $(version $App) ]] &&
{	mv "$App" ~/.Trash
	mv "$1" "${App%/*}"
	ln -s "$App" "${1%/*}"/
}
```

**Comment by brianroach** [comment]
Thanks @danielbayley!
